overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,785 @@
+package common
+
+import (
+	"encoding/base32"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	ErrInvalidSnapLabel           = errors.New("the given label cannot be converted to snap")
+	ErrNoPatterns                 = errors.New("no patterns given, cannot establish precedence")
+	ErrPermissionNotInList        = errors.New("permission not found in permissions list")
+	ErrPermissionsListEmpty       = errors.New("permissions list empty")
+	ErrUnrecognizedFilePermission = errors.New("file permissions mask contains unrecognized permission")
+)
+
+type Constraints struct {
+	PathPattern string   `json:"path-pattern"`
+	Permissions []string `json:"permissions"`
+}
+
+func (constraints *Constraints) ValidateForInterface(iface string) error {
+	switch iface {
+	case "home", "camera":
+		if err := ValidatePathPattern(constraints.PathPattern); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("constraints incompatible with the given interface: %s", iface)
+	}
+	permissions, err := AbstractPermissionsFromList(iface, constraints.Permissions)
+	if err != nil {
+		return err
+	}
+	constraints.Permissions = permissions
+	return nil
+}
+
+func (constraints *Constraints) Match(path string) (bool, error) {
+	return PathPatternMatch(constraints.PathPattern, path)
+}
+
+// Removes the given permission from the permissions associated with the
+// constraints. Assumes that the permission occurs at most once in the list.
+// If the permission does not exist in the list, returns ErrPermissionNotInList.
+func (constraints *Constraints) RemovePermission(permission string) error {
+	origLen := len(constraints.Permissions)
+	i := 0
+	for i < len(constraints.Permissions) {
+		perm := constraints.Permissions[i]
+		if perm != permission {
+			i++
+			continue
+		}
+		copy(constraints.Permissions[i:], constraints.Permissions[i+1:])
```

Comment:
Perhaps we can use something from the slices package: https://pkg.go.dev/slices#DeleteFunc